File: Warm-up Challanges (HackRank)/Counting Valleys/Counting Valleys.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

/*
 * Complete the 'countingValleys' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER steps
 *  2. STRING path
 */

function countingValleys(steps, path) {
    // Write your code here
    let result=0;
    let valley=0;
    let canSum=false;
    for(let i = 0; steps > i; i++) {
        if(path[i] === 'U') {
            valley += 1;
        } 
        
        if(path[i] === 'D') {
            valley -= 1;
        }
        
        if(valley<0) {
            canSum = true;
        }
        
        if(valley === 0 && canSum) {
            canSum = false;
            result ++;
        }
    }
    return result;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const steps = parseInt(readLine().trim(), 10);

    const path = readLine();

    const result = countingValleys(steps, path);

    ws.write(result + '\n');

    ws.end();
}

File: Warm-up Challanges (HackerRank)/Counting Valleys/Counting Valleys.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

class Result {

    /*
     * Complete the 'countingValleys' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts following parameters:
     *  1. INTEGER steps
     *  2. STRING path
     */

    public static int countingValleys(int steps, String path) {
    // Write your code here
        int result = 0;
        int valley = 0;
        boolean canSum = false;
        
        for(int i = 0; steps > i; i++) {
            String currentPath = Character.toString(path.charAt(i));

            if(currentPath.equals("U")) {
                valley++;
            }
            
            if(currentPath.equals("D")) {
                valley--;
            }
            
            if(valley < 0){
                canSum = true;
            }
            
            if(valley == 0 && canSum) {
                canSum = false;
                result++;
            }
        }
        return result++;
    }

}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int steps = Integer.parseInt(bufferedReader.readLine().trim());

        String path = bufferedReader.readLine();

        int result = Result.countingValleys(steps, path);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}
